File: services/api/app/middleware/security_headers.py
from __future__ import annotations

from collections.abc import Awaitable, Callable

from fastapi import Request, Response
from starlette.middleware.base import BaseHTTPMiddleware

from app.settings import Settings, get_settings

HSTS_VALUE = "max-age=63072000; includeSubDomains; preload"
API_CSP = "default-src 'none'; frame-ancestors 'none'; base-uri 'none'; form-action 'none'"
PERMISSIONS_POLICY = (
    "camera=(), microphone=(), geolocation=(), browsing-topics=(), payment=(), usb=()"
)

# Browser-facing CORS preflight allowlist (webhooks / n8n use server-to-server headers).
CORS_ALLOW_METHODS = ["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"]
CORS_ALLOW_HEADERS = [
    "Authorization",
    "Content-Type",
    "Accept",
    "X-Request-ID",
    "Idempotency-Key",
]


class SecureHeadersMiddleware(BaseHTTPMiddleware):
    """Baseline security headers for JSON API responses (mirrors infra/Caddyfile api_security)."""

    async def dispatch(
        self,
        request: Request,
        call_next: Callable[[Request], Awaitable[Response]],
    ) -> Response:
        settings = get_settings()
        response = await call_next(request)
        _apply_secure_headers(response, settings)
        return response


def _apply_secure_headers(response: Response, settings: Settings) -> None:
    response.headers["X-Content-Type-Options"] = "nosniff"
    response.headers["X-Frame-Options"] = "DENY"
    response.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"
    response.headers["Content-Security-Policy"] = API_CSP
    response.headers["Permissions-Policy"] = PERMISSIONS_POLICY
    response.headers["Cross-Origin-Opener-Policy"] = "same-origin"
    if settings.env != "development":
        response.headers["Strict-Transport-Security"] = HSTS_VALUE

File: packages/config/src/security-headers.ts
/**
 * Shared security-header constants and CSP helpers (M15-P03 + security audit wave).
 *
 * Used by Next.js `next.config.ts` (build-time headers) and `middleware.ts`
 * (per-request report-only CSP). Keeps API / CDN / Vercel allowlists in one place.
 */

export type EnvBag = Record<string, string | undefined>;

export const HSTS_VALUE = "max-age=63072000; includeSubDomains; preload";

export const CSP_ORIGINS = {
  cloudinary: "https://res.cloudinary.com",
  supabase: "https://*.supabase.co",
  supabaseWs: "wss://*.supabase.co",
  ga4Script: "https://*.googletagmanager.com",
  ga4Connect:
    "https://*.google-analytics.com https://*.analytics.google.com https://*.googletagmanager.com",
  ga4Img: "https://*.google-analytics.com https://*.googletagmanager.com",
  lencoWidget: "https://pay.lenco.co https://pay.sandbox.lenco.co",
  lencoApi: "https://api.lenco.co https://api.sandbox.lenco.co",
  sentryIngest:
    "https://*.ingest.sentry.io https://*.ingest.us.sentry.io https://*.ingest.de.sentry.io",
  vercel: "https://*.vercel.app https://vitals.vercel-insights.com https://vercel.live",
  apiProduction: "https://api.vergeo5.com",
  apiStaging: "https://api.staging.vergeo5.com",
} as const;

export const PERMISSIONS_POLICY_DEFAULT =
  "camera=(), microphone=(), geolocation=(), browsing-topics=(), payment=(), usb=()";

export const PERMISSIONS_POLICY_ADMIN =
  "camera=(), microphone=(), geolocation=(), browsing-topics=(), payment=(), usb=(), " +
  "accelerometer=(), gyroscope=(), magnetometer=(), fullscreen=(self)";

export function isDevelopmentEnv(env: EnvBag = process.env): boolean {
  return env.NODE_ENV !== "production";
}

/**
 * Origins the browser may `fetch()` against (cross-origin API + local dev).
 * Includes configured `NEXT_PUBLIC_API_BASE_URL` / `NEXT_PUBLIC_VERGEO_API_URL`
 * plus known prod/staging API hosts so preview builds stay functional.
 */
export function resolveApiConnectOrigins(env: EnvBag = process.env): string {
  const origins = new Set<string>([CSP_ORIGINS.apiProduction, CSP_ORIGINS.apiStaging]);

  for (const key of ["NEXT_PUBLIC_API_BASE_URL", "NEXT_PUBLIC_VERGEO_API_URL"] as const) {
    const trimmed = env[key]?.trim();
    if (trimmed) {
      origins.add(trimmed.replace(/\/$/, ""));
    }
  }

  // Direct `process.env.NODE_ENV` comparison so production Next.js builds DCE
  // the loopback literal out of middleware / CSP helpers. Passing `env` into
  // `isDevelopmentEnv` is not enough for Terser to drop the string.
  if (process.env.NODE_ENV !== "production") {
    if (isDevelopmentEnv(env)) {
      origins.add("http://localhost:8000");
    }
  }

  return [...origins].join(" ");
}

export function buildConnectSrc(env: EnvBag = process.env, extra = ""): string {
  const api = resolveApiConnectOrigins(env);
  const vercel = isDevelopmentEnv(env) ? "" : ` ${CSP_ORIGINS.vercel}`;
  const suffix = extra ? ` ${extra}` : "";
  return `'self' ${api} ${CSP_ORIGINS.supabase} ${CSP_ORIGINS.supabaseWs}${vercel}${suffix}`
    .replace(/\s+/g, " ")
    .trim();
}

export type StaticSecurityHeader = { key: string; value: string };

/** Baseline headers applied on every route (HSTS omitted in development). */
export function buildStaticSecurityHeaders(options: {
  xFrameOptions: "DENY" | "SAMEORIGIN";
  permissionsPolicy: string;
  enforcedCsp: string;
  env?: EnvBag;
}): StaticSecurityHeader[] {
  const env = options.env ?? process.env;
  const headers: StaticSecurityHeader[] = [];

  if (!isDevelopmentEnv(env)) {
    headers.push({ key: "Strict-Transport-Security", value: HSTS_VALUE });
  }

  headers.push(
    { key: "X-Content-Type-Options", value: "nosniff" },
    { key: "X-Frame-Options", value: options.xFrameOptions },
    { key: "Referrer-Policy", value: "strict-origin-when-cross-origin" },
    { key: "Permissions-Policy", value: options.permissionsPolicy },
    { key: "Cross-Origin-Opener-Policy", value: "same-origin" },
    { key: "Content-Security-Policy", value: options.enforcedCsp },
  );

  return headers;
}
